Added API ceagle proxy to health

teferi opened the pull request.

teferi committed:
Implement health API proxy for ceagle

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -13,15 +13,34 @@
 bp_region_status = flask.Blueprint("region", __name__)
 
 
+def get_endpoint(service):
```
API client already cares about endpoint
https://github.com/seecloud/ceagle/blob/c749cf1d1a148ed4ca12ed409febea42b48b71c8/ceagle/api/client.py#L56

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -13,15 +13,34 @@
+import json
+
 import flask
+import requests
```
No need to use requests directly, there is an API client for that
https://github.com/seecloud/ceagle/blob/c749cf1d1a148ed4ca12ed409febea42b48b71c8/ceagle/api/client.py#L56

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -55,7 +84,17 @@ def get_region_status(region, period):
+    health_endpoint = get_endpoint("health")
+    api_endpoint = "/api/v1/health/{}".format(region)
+    health_endpoint + api_endpoint
+    resp = requests.get(
```
Use API client

boris-42 commented:
I would prefer to see in commit: "Implement Health API"

teferi committed:
Modify api_fake_data's api_handler to be lazy

Previously api_handler function returned a lambda, that could be used as
a decorator and would return fake or real function, depending on a
global variable. The decision was made at import time. This made it very
difficult to test decorated functions behaviour, since these functions
were decorated away before there was any chance to mock config or global
variable.
With this commit api_handler returns a decorator, that, when applied,
returns a function, that proxies calls to fake or real api function,
depending on global variable.
This consequently means there will be an extra function call for every
request. However this would potentially allow us to switch behaviour
from/to fake functions on the fly.

This commit also restores fake decorators around health API functions.

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -29,9 +32,17 @@ def get_status(period):
+    if not health_client:
+        return flask.make_response(
+            json.dumps({"error": "No health endpoint configured"}),
+            500,
```
This is definitely not "500" since 500 means unexpected server error, however this case is a normal workflow.
I would not even think this is 503 because 503 is "service must exist but absent now".

I would recommend to use "404 Not Found" here because this is actually a client problem - we do not have health service configured (so it must not be provided) however client tries to access this absent entity.

teferi replied:
500 does not mean "unexpected server error", but rather internal server error. https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html 

>    The server encountered an unexpected condition which prevented it
>    from fulfilling the request.

That's exactly what we have. The server is not configured properly to handle this request. 404 is not applicable here (again refer to the rfc):
> The server has not found anything matching the Request-URI

We've found all we have, but the server is misconfigured.

503 is definitely not applicable, see rfc
502 might be a better choice, but since the upstream is not configured at all — 500 is the best status here.

maretskiy replied:
404 because client requests service which is absent for this specific region

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -29,9 +32,17 @@ def get_status(period):
+            500,
+        )
+
+    api_endpoint = "/api/v1/health"
```
Endpoint comes from config, do not hard-code such data
https://github.com/seecloud/ceagle/blob/297b03a51ed0108b60088c5626ad7d0f25cb3299/etc/eagle_config.json#L10

teferi replied:
this is API endpoint of health api and it should be hardcoded here, since there is no way to devise it from just uri, where api runs

teferi replied:
for example health api has '/api/v1/regions' endpoint — if you put "http://url:port/api/v1/health" in the health endpoint — there is no way to get regions api endpoint from health.

maretskiy replied:
Ok, agree

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -53,9 +64,17 @@ def get_region_status(region, period):
+            500,
+        )
+
+    api_endpoint = "/api/v1/health/{}".format(region)
```
Same here, endpoint part comes from config

uri = "/{}".format(region)
result = health_client.get(uri, params={"period": period})

teferi replied:
see prev comment

maretskiy commented on tests/unit/api/v1/test_status.py:
```diff
@@ -13,19 +13,19 @@
-from tests.unit import test
+import mock
+
+from tests.unit import test  # noqa
```
"# noqa" should be removed. There was a temporarily workarond when there was an issue with tests, not everything should work fine without "# noqa"

maretskiy replied:
*now everything should work fine without "# noqa"

teferi replied:
ok, although I see little harm in having it here.

maretskiy commented on tests/unit/api/v1/test_status.py:
```diff
@@ -38,3 +38,47 @@ def test_api_response_code(self):
+
+    def setUp(self):
+        super(HealthApiTestCase, self).setUp()
+        self.health_config = {"health": {
```
It is better to pre-configure mock_get_config for every test method - this will really save LOC
https://docs.python.org/3.5/library/unittest.mock-examples.html#applying-the-same-patch-to-every-test-method

maretskiy commented on ceagle/api/v1/status.py:
```diff
@@ -29,9 +32,17 @@ def get_status(period):
 
 
 @bp_status.route("/health/<period>")
-@fake_status.get_status_health
```
I would recommend to keep this decorator but apply changes related to this decorator as first commit, so both real and fake code could be tested normally in unit tests in this commit

teferi replied:
this decorator is restored in the very same pull request — see 2d commit

maretskiy commented on tests/unit/api/v1/test_status.py:
```diff
@@ -45,9 +47,18 @@ class HealthApiTestCase(test.TestCase):
+        }
+
+        self.old_USE_FAKE_DATA = fake_api_base.USE_FAKE_DATA
+        fake_api_base.USE_FAKE_DATA = False
```
This is not clear to me, does this mean that Fake API is not tested?

What do you think about approach of having 2 test classes for each module:

class ApiTestCase(test.TestCase):
     <tests for real API>

class FakeApiTestCase(test.TestCase):
     <tests for fake API>

teferi replied:
Fake API is tested in ApiTestCases. I've restored all the lines deleted from there.

We can rename them as soon as there would be additional cases.

maretskiy commented on tests/unit/api/v1/test_status.py:
```diff
@@ -38,3 +38,47 @@ def test_api_response_code(self):
+
+import unittest
+
+if __name__ == '__main__':
```
Why this is needed?

teferi replied:
I largely forgot to remove this one =)
